src/libappimage/utils/ElfFile.cpp

```diff
@@ -0,0 +1,151 @@
+
+            fseeko(fd, 0, SEEK_SET);
+            ret = fread(&ehdr32, 1, sizeof(ehdr32), fd);
+            if (ret < 0 || (size_t) ret != sizeof(ehdr32)) {
+                fprintf(stderr, "Read of ELF header from %s failed: %s\n",
+                        fname, strerror(errno));
+                return -1;
+            }
+
+            ehdr.e_shoff = file32_to_cpu(ehdr32.e_shoff);
+            ehdr.e_shentsize = file16_to_cpu(ehdr32.e_shentsize);
+            ehdr.e_shnum = file16_to_cpu(ehdr32.e_shnum);
+
+            last_shdr_offset = ehdr.e_shoff + (ehdr.e_shentsize * (ehdr.e_shnum - 1));
+            fseeko(fd, last_shdr_offset, SEEK_SET);
+            ret = fread(&shdr32, 1, sizeof(shdr32), fd);
+            if (ret < 0 || (size_t) ret != sizeof(shdr32)) {
+                fprintf(stderr, "Read of ELF section header from %s failed: %s\n",
+                        fname, strerror(errno));
+                return -1;
+            }
+
+            /* ELF ends either with the table of section headers (SHT) or with a section. */
+            sht_end = ehdr.e_shoff + (ehdr.e_shentsize * ehdr.e_shnum);
+            last_section_end = file64_to_cpu(shdr32.sh_offset) + file64_to_cpu(shdr32.sh_size);
+            return sht_end > last_section_end ? sht_end : last_section_end;
+        }
+
+        off_t ElfFile::read_elf64(FILE* fd) {
+            Elf64_Ehdr ehdr64;
+            Elf64_Shdr shdr64;
+            off_t last_shdr_offset;
+            off_t ret;
+            off_t sht_end, last_section_end;
+
+            fseeko(fd, 0, SEEK_SET);
+            ret = fread(&ehdr64, 1, sizeof(ehdr64), fd);
+            if (ret < 0 || (size_t) ret != sizeof(ehdr64)) {
+                fprintf(stderr, "Read of ELF header from %s failed: %s\n",
+                        fname, strerror(errno));
+                return -1;
+            }
+
+            ehdr.e_shoff = file64_to_cpu(ehdr64.e_shoff);
+            ehdr.e_shentsize = file16_to_cpu(ehdr64.e_shentsize);
+            ehdr.e_shnum = file16_to_cpu(ehdr64.e_shnum);
+
+            last_shdr_offset = ehdr.e_shoff + (ehdr.e_shentsize * (ehdr.e_shnum - 1));
+            fseeko(fd, last_shdr_offset, SEEK_SET);
+            ret = fread(&shdr64, 1, sizeof(shdr64), fd);
+            if (ret < 0 || ret != sizeof(shdr64)) {
+                fprintf(stderr, "Read of ELF section header from %s failed: %s\n",
+                        fname, strerror(errno));
+                return -1;
+            }
+
+            /* ELF ends either with the table of section headers (SHT) or with a section. */
+            sht_end = ehdr.e_shoff + (ehdr.e_shentsize * ehdr.e_shnum);
+            last_section_end = file64_to_cpu(shdr64.sh_offset) + file64_to_cpu(shdr64.sh_size);
+            return sht_end > last_section_end ? sht_end : last_section_end;
+        }
+
+        ssize_t ElfFile::getSize() {
+            off_t ret;
+            FILE* fd = NULL;
+            off_t size = -1;
+
+            fd = fopen(fname, "rb");
```

Comment:
Shouldn't use any `fprintf` _ever_ in the codebase.

`fd` etc. could be managed more easily by a shared pointer, just pass `fclose` as deleter.

Reply:
Old legacy code, replacing by the logger.